alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/binder/solr/sharding/SolrShardingMetricsContainer.java

```diff
@@ -45,20 +47,25 @@ public void refresh() {
     }
 
     public Set<ShardInstance> getShardInstances(Shard shard) {
-        return rawData.get(shard.getFloc()).get(shard).stream().map(shardState -> shardState.getShardInstance())
+        return rawData.get(shard.getFloc()).get(shard).stream().filter(shardState -> Objects.nonNull(shardState))
+                .map(shardState -> shardState.getShardInstance()).filter(instance -> instance != null)
                 .collect(
                         Collectors.toSet());
     }
 
-    public ShardState getShardState(ShardInstance shardInstance) {
+    public Optional<ShardState> getShardState(ShardInstance shardInstance) {
         Shard shard = shardInstance.getShard();
         return rawData.get(shard.getFloc()).get(shard).stream()
-                .filter(shardState -> shardState.getShardInstance() == shardInstance).findAny().get();
+                .filter(shardState -> shardState.getShardInstance() == shardInstance).findAny();
     }
 
-    public ReplicaState getReplicaState(ShardInstance shardInstance) {
-        return ReplicaState
-                .valueOf(getShardState(shardInstance).getPropertyBag().get(ShardRegistryImpl.INSTANCE_STATE));
+    public Optional<ReplicaState> getReplicaState(ShardInstance shardInstance) {
+        Optional<ShardState> shardState = getShardState(shardInstance);
+        if (!shardState.isPresent()) {
```

Comment:
Instead of doing an `isPresent` check, you can use [`Optional.flatMap()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#flatMap-java.util.function.Function-) or [`Optional.map()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-), depending on what you need.

I think `Optional.map()` is enough here, since you are using `Optional.of()` below, which does not accept nullable values.

Reply:
Hey @vierbergenlars .

Do you mean like this for example:

`
  return Optional.of(ReplicaState.valueOf(shardState.filter(shardstate -> Objects.nonNull(shardState))
                .get().getPropertyBag().get(ShardRegistryImpl.INSTANCE_STATE)));
`

I don't really get how I can use shardState.map() for this.

Reply:
Okey I figured it out

` return getShardState(shardInstance)
                .map(shardState1 -> ReplicaState.valueOf(shardState1.getPropertyBag().get(ShardRegistryImpl.INSTANCE_STATE)));`